### ISSUE TITLE
Failed to delete data from log stream

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3562]
**Describe the bug**
It seems even after the fix in https://github.com/zeebe-io/zeebe/pull/3522, we are some times not able to delete data. It happens that sometimes the following is logged `Tried to delete from log stream, but found no corresponding address for the given position 172902503404768.
`. But might be not a real issue since we eventually should be able to delete the log, on later snapshot.

**To Reproduce**

Just run a benchmark and wait.

**Expected behavior**
Delete the log and no problems.

**Log/Stacktrace**
I run a benchmark with three nodes and one partition. On the follower we saw this log statement.

<details><summary>Full Stacktrace</summary>
 <pre>
I 2019-12-17T04:30:52.623912246Z 2019-12-17 04:30:52.623 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-1.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams - Delete data from log stream until position '172902503404768' (address: '73042331').
 
I 2019-12-17T04:30:52.624080761Z 2019-12-17 04:30:52.623 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 71343303 up to 72973592
 
I 2019-12-17T04:30:52.624245895Z 2019-12-17 04:30:52.624 [] [raft-partition-group-raft-partition-6] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 72973592
 
I 2019-12-17T04:30:54.941545602Z 2019-12-17 04:30:54.907 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-0.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-2] DEBUG io.zeebe.logstreams - Delete data from log stream until position '172902503404768' (address: '73042331').
 
I 2019-12-17T04:30:54.942372062Z 2019-12-17 04:30:54.942 [] [raft-server-0-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 71343303 up to 72973592
 
I 2019-12-17T04:30:54.942742228Z 2019-12-17 04:30:54.942 [] [raft-partition-group-raft-partition-6] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 72973592
 
I 2019-12-17T04:30:55.877582974Z 2019-12-17 04:30:55.877 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-2.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.logstreams - Tried to delete from log stream, but found no corresponding address for the given position 172902503404768.
 
</pre>
</details>

**Environment:**
 - OS: k8s
 - Zeebe Version: SNAPSHOT
 - Configuration: no exporter

**Comment by npepinpe** [comment]
Did this happen across restarts? I think it might be the case we discussed, where if you have gaps in your log and you receive a snapshot it might point to an entry that you don't have (and the leader won't send either).

**Comment by npepinpe** [comment]
So the idea here is that snapshots already have an index associated, and we should be able to reuse those for easy, simple deletion. The problem is the log stream itself does not know about "indices", rather it just knows about positions. So conceptually it's iffy to have `LogStream#delete(long)` take in an index instead of a position...but it is a lot simpler than finding the correct entry for the given position and handling gaps.

The naive solution here is to have the `LogStreamDeletionService` become the `LogStorageDeletionService` and have a reference to the `LogStorage` instead of the `LogStream`, such that we just immediately call `LogStorage#delete(long address)` with the snapshot index. What do you think, @Zelldon?

**Comment by Zelldon** [comment]
Right out my head; Lets not directly break our API which we just cleaned up :sweat_smile: I would vote for the second option then, which is then also more portable

**Comment by Zelldon** [comment]
Could we just unify these position index thing such that is used everywhere. Like writing to ring buffer creates new index (previous + 1) and this is also the index in the log? Maybe it then also makes sense to map one indexed entry to one record? Would this make the things easier?

**Comment by npepinpe** [comment]
It would definitely make it easier, but that's a bigger issue, since the Raft itself sometimes write entries without going through the ring buffer, so we'd somehow need to keep them in sync...I'm also not sure how we would batch records into a single entry :thinking:

**Comment by npepinpe** [comment]
Let's talk about it tomorrow, I think maybe I'm approaching this the wrong way; this is only an issue on catch up, so maybe this is something we have to handle for that particular case.

**Comment by Zelldon** [comment]
Seems to be an bigger issue, because I have two benchmarks which fail after these lines are printed.
```
Tried to delete from log stream, but found no corresponding address for the given position 277334634514816.
```

On the benchmark with the snapshot version, these lines are printed as last:
```
2019-12-17 14:50:10.758 CET
2019-12-17 13:50:10.758 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-2.zeebe.one-partition-throughput.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.logstreams - Tried to delete from log stream, but found no corresponding address for the given position 277334634514816.
 
2019-12-17 15:04:45.028 CET
2019-12-17 14:04:45.028 [partition-1-processor-snapshot-director] [zeebe-0.zeebe.one-partition-throughput.svc.cluster.local:26501-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - Take temporary snapshot and write into /usr/local/zeebe/data/raft-partition/partitions/1/pending/117597322-1-1576591485028-280109180096472.
2019-12-17 15:04:45.207 CET
2019-12-17 14:04:45.207 [partition-1-processor-snapshot-director] [zeebe-0.zeebe.one-partition-throughput.svc.cluster.local:26501-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - Creation of snapshot for partition-1-processor took 10463 ms.
2019-12-17 15:04:45.208 CET
2019-12-17 14:04:45.207 [partition-1-processor-snapshot-director] [zeebe-0.zeebe.one-partition-throughput.svc.cluster.local:26501-zb-fs-workers-0] INFO io.zeebe.logstreams.snapshot - Finished taking snapshot, need to wait until last written event position 280109180096472 is committed, current commit position is 280104888573240. After that snapshot can be marked as valid.
```

On the other benchmark where I tested the transport removing this are the last log statements:

```

2019-12-17 12:45:47.266 CET
2019-12-17 11:45:47.266 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-2.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.logstreams - Tried to delete from log stream, but found no corresponding address for the given position 225679058541696.
2019-12-17 12:45:49.078 CET
2019-12-17 11:45:49.075 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-1.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-2] DEBUG io.zeebe.logstreams - Delete data from log stream until position '225679058541696' (address: '94950497').
2019-12-17 12:45:49.078 CET
2019-12-17 11:45:49.078 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 94108600 up to 94916865
2019-12-17 12:45:49.079 CET
2019-12-17 11:45:49.078 [] [raft-partition-group-raft-partition-13] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 94916865
2019-12-17 12:45:53.126 CET
2019-12-17 11:45:53.113 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-0.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams - Delete data from log stream until position '225679058541696' (address: '94950497').
2019-12-17 12:45:53.133 CET
2019-12-17 11:45:53.132 [] [raft-server-0-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 94108600 up to 94916865
2019-12-17 12:45:53.134 CET
2019-12-17 11:45:53.133 [] [raft-partition-group-raft-partition-13] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 94916865
2019-12-17 13:00:41.748 CET
2019-12-17 12:00:41.748 [io.zeebe.broker.logstreams.LogStreamDeletionService] [zeebe-2.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - The last processed position for partition 1 is 227530190775040
2019-12-17 13:00:41.867 CET
2019-12-17 12:00:41.867 [io.zeebe.broker.logstreams.LogStreamDeletionService] [zeebe-1.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-3] DEBUG io.zeebe.broker.system - The last processed position for partition 1 is 227530190775040
2019-12-17 13:00:41.968 CET
2019-12-17 12:00:41.967 [io.zeebe.broker.logstreams.LogStreamDeletionService] [zeebe-0.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.broker.system - The last processed position for partition 1 is 227530190775040
2019-12-17 13:00:46.243 CET
2019-12-17 12:00:46.243 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-2.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams - Tried to delete from log stream, but found no corresponding address for the given position 227530190775040.
2019-12-17 13:00:48.299 CET
2019-12-17 12:00:48.290 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-1.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-2] DEBUG io.zeebe.logstreams - Delete data from log stream until position '227530190775040' (address: '95715637').
2019-12-17 13:00:48.299 CET
2019-12-17 12:00:48.299 [] [raft-server-1-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 94916865 up to 95609061
2019-12-17 13:00:48.299 CET
2019-12-17 12:00:48.299 [] [raft-partition-group-raft-partition-3] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 95609061
2019-12-17 13:00:53.653 CET
2019-12-17 12:00:53.632 [io.zeebe.logstreams.impl.log.LogStreamImpl] [zeebe-0.zeebe.zell-remove-transport.svc.cluster.local:26501-zb-actors-3] DEBUG io.zeebe.logstreams - Delete data from log stream until position '227530190775040' (address: '95715637').
2019-12-17 13:00:53.653 CET
2019-12-17 12:00:53.653 [] [raft-server-0-raft-partition-partition-1] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting log up from 94916865 up to 95609061
2019-12-17 13:00:53.654 CET
2019-12-17 12:00:53.653 [] [raft-partition-group-raft-partition-3] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 95609061
2019-12-17 13:01:53.799 CET
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:159) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?] at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?] at java.lang.Thread.run(Unknown Source) ~[?:?] Caused by: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds ... 7 more
2019-12-17 13:01:53.799 CET
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 59997 milliseconds at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$6(NettyMessagingService.java:329) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$executeOnPooledConnection$7(NettyMessagingService.java:329) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:159) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?] at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?] at java.lang.Thread.run(Unknown Source) ~[?:?] Caused by: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 59997 milliseconds ... 7 more
2019-12-17 13:01:53.799 CET
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:159) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?] at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?] at java.lang.Thread.run(Unknown Source) ~[?:?] Caused by: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds ... 7 more
2019-12-17 13:01:53.800 CET
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?] at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?] at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?] at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.timeout(AbstractClientConnection.java:159) ~[atomix-cluster-3.2.0-20191207.093027-100.jar:?] at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?] at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?] at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?] at java.lang.Thread.run(Unknown Source) ~[?:?] Caused by: java.util.concurrent.TimeoutException: Request type command-api-1 timed out in 60000 milliseconds ... 7 more
```

So in both cases it was on broker 2, which is a follower, where the log can't be deleted. But this seems to be really problematic because after that not longer processing happens. But the disk is not full !


|    | SNAPSHOT | remove transport |
|--------|---------|------|
|processing|![snapshot-throughput](https://user-images.githubusercontent.com/2758593/71067844-98670500-2175-11ea-9473-02099dd8ad2d.png)|![remove-transport](https://user-images.githubusercontent.com/2758593/71067767-6bb2ed80-2175-11ea-8aab-7eefee943b19.png)|
|resources|![snapshot-resources](https://user-images.githubusercontent.com/2758593/71067788-79687300-2175-11ea-8a56-2a57be3aec1c.png)|![remove-transport-resources](https://user-images.githubusercontent.com/2758593/71067786-79687300-2175-11ea-9b51-13f79d3ae837.png)|

**Comment by Zelldon** [comment]
This only happens btw with one partition, on multiple partition setup I never saw this issue. 
Furthermore it is interesting that I saw before the not able to delete log statement repeatedly the following on both benchmarks


```
I 2019-12-17T13:50:03.775623143Z 2019-12-17 13:50:03.773 [] [raft-server-2-raft-partition-partition-1] WARN  io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot - Failed to delete pending snapshot DbPendingSnapshot{index=117067677, term=1, timestamp=2019-12-17 01:34:46,371, directory=/usr/local/zeebe/data/raft-partition/partitions/1/pending/117067677-1-1576589686371}
 
I 2019-12-17T13:50:03.775709705Z java.nio.file.NoSuchFileException: /usr/local/zeebe/data/raft-partition/partitions/1/pending/117067677-1-1576589686371
	at sun.nio.fs.UnixException.translateToIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.LinuxFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.Files.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.getAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.visit(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.walk(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at io.zeebe.util.FileUtil.deleteFolder(FileUtil.java:41) ~[zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot.abort(DbPendingSnapshot.java:120) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.atomix.protocols.raft.roles.PassiveRole.abortPendingSnapshot(PassiveRole.java:503) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.roles.PassiveRole.onInstall(PassiveRole.java:242) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.roles.FollowerRole.onInstall(FollowerRole.java:80) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.lambda$null$10(RaftContext.java:217) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.lambda$runOnContext$33(RaftContext.java:254) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-20191207.093016-100.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
 
I 2019-12-17T13:50:03.811444004Z 2019-12-17 13:50:03.810 [partition-1-processor-snapshot-director] [zeebe-0.zeebe.one-partition-throughput.svc.cluster.local:26501-zb-fs-workers-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Search for orphaned snapshots below oldest valid snapshot DbSnapshotMetadata{index=116437912, term=1, timestamp=2019-12-17 01:20:02,175, position=277291684030912}
 
```

**Comment by Zelldon** [comment]
@npepinpe what do you think of introducing the previous read address. This means when seek was successful, we use the last read address and when seek failed we use the previous read address, which contains then the previous index with smaller positions. I think we have two cases where seek can fail, either the next position is greater then the searched one then we could use the previous read address. Or we reached the end of the log and haven't found the position and no greater position. We need to differentiate these cases then we can handle them also separately.

**Comment by npepinpe** [comment]
I think we used to have this and removed it because it was very much a hack. We could look into it again; I think with binary search this is already the case, its just discarded by the log stream reader implementation.

But I think we have bigger problems than this. I think our snapshots are broken; the normal Raft algorithm (or at least our implementation) doesn't take into accounts multiple snapshots, so right now we might end up with gaps in our logs, or at least snapshots which should not be recoverable.

Take for example two nodes, L and F, with the following logs:

L: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
F: 1, 2, 3, 4

L has snapshots S_7, S_8, S_9 (pointing a 7, 8, and 9 respectively), and F has snapshots S_1, S_2, S_3.

There are two cases where L will try to send a snapshot to F:

1. F was gone and is catching up
1. F is too slow to replicate and keep up

In either case, L will always send its __latest__ snapshot. So when F rejoins (or tries to catch up), it now has snapshots S_1, S_2, S_3, S_9. It will delete S_1 and compact properly, no problem. It will also receive event 10. The problem is that the snapshots S_2 and S_3 are now useless - if we were to recover from them on F (since F can now be leader), we could, but then we'd go from event 4 to event 10 without realizing, resulting in an inconsistent log. This can happen if for whatever reason S_9 is corrupted on F (the whole reason we keep 3 snapshots).

**Comment by Zelldon** [comment]
This sounds bad. But maybe we can fix this separately. So I would propose that we go with the solution I wrote above, to fix the deletion problem quickly and then take a look how we can solve the issue with inconsistent log or gaps in the log and a leader change. What do you say?

Would it help if we send all three snapshots? Or maybe on receiving check whether this snapshot is corrupted, if this is the case we request a new one?

**Comment by npepinpe** [comment]
I think you're right and these are two separate issues (though related). I'll make another issue and we can discuss it there.

**Comment by npepinpe** [comment]
Regarding this issue, let's discuss if we really need to keep this compaction stuff portable - right now we decided so because we consider this part of the log stream abstraction, but I'm not sure that is the right scope; after all, compaction/snapshotting is part of the Raft paper, and I'm not sure adding that to our log stream abstraction is useful or necessary. The way I see it, the log stream is meant to provide an abstraction layer over a stream of records - it doesn't care about replication, persistence, size, etc. That's why we have an abstraction.

But I could be wrong, so I'd like to discuss that.

**Comment by Zelldon** [comment]
:+1: Ok lets discuss this f2f

**Comment by npepinpe** [comment]
Closed by #3962
